cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,481 @@
+    //which won't be called anymore (because of the _condition_ below)
+    //
+    cudf::set_null_mask(merged_view.null_mask(),
+                        merged_view.size(),
+                        ALL_VALID,
+                        stream_);
+
+    //to resolve view.data()'s types use: ElemenT
+    //
+    ElemenT const* p_d_lcol = lcol.data<ElemenT>();
+    ElemenT const* p_d_rcol = rcol.data<ElemenT>();
+        
+    auto exe_pol = rmm::exec_policy(stream_);
+    
+    //capture lcol, rcol
+    //and "gather" into merged_view.data()[indx_merged]
+    //from lcol or rcol, depending on side;
+    //
+    thrust::transform(exe_pol->on(stream_),
+                      dv_row_order_.begin(), dv_row_order_.end(),
+                      merged_view.begin<ElemenT>(),
+                      [p_d_lcol, p_d_rcol] __device__ (index_type const& index_pair){
+                       auto side = thrust::get<0>(index_pair);
+                       auto index = thrust::get<1>(index_pair);
+                       
+                       ElemenT val = (side == side::LEFT ? p_d_lcol[index] : p_d_rcol[index]);
+                       return val;
+                      }
+                     );
+
+    //cudaDeviceSynchronize();//? nope...these two could proceed concurrently
+
+
+    //CAVEAT: conditional call below is erroneous without
+    //set_null_mask() call (see TODO above):
+    //
+    if (lcol.has_nulls() || rcol.has_nulls())
+      //resolve null mask:
+      //
+      materialize_bitmask(lcol,
+                          rcol,
+                          merged_view,
+                          dv_row_order_.data().get(),
+                          stream_);
+                   
+    return p_merged_col;
+  }
+
+  //specialization for string...?
+  //or should use `cudf::string_view` instead?
+  //
+  template<typename ElemenT>//required: column type
+  std::enable_if_t<not cudf::is_fixed_width<ElemenT>(),
+                   std::unique_ptr<cudf::column>>
+  operator()(cudf::column_view const& lcol, cudf::column_view const& rcol) const
+  {
+    //for now...
+    CUDF_FAIL("Non fixed-width types are not supported");
+  }
+
+private:
+  VectorI const& dv_row_order_;
+  rmm::mr::device_memory_resource* mr_;
+  cudaStream_t stream_;
+  
+  //see `class element_relational_comparator` in `cpp/include/cudf/table/row_operators.cuh` as a model;
+};
+  
+
+  std::unique_ptr<cudf::experimental::table> merge(cudf::table_view const& left_table,
+                                                   cudf::table_view const& right_table,
+                                                   std::vector<cudf::size_type> const& key_cols,
+                                                   std::vector<cudf::order> const& column_order,
+                                                   std::vector<cudf::null_order> const& null_precedence) {
+    auto n_cols = left_table.num_columns();
+    CUDF_EXPECTS( n_cols == right_table.num_columns(), "Mismatched number of columns");
```

Comment:
I don't understand. How do you merge tables with different number of columns? How many columns will there be in the resulting table? What do you fill the missing columns with?

Reply:
Yeah, I wasn't thinking right. Conflated this with join. Disregard. 

There should still be `left_key_columns` and `right_key_columns` however.

Reply:
If left and right key columns are in different positions then you have the same column miss-alignment problem that you would get by having different number or types of columns. 

The algorithm would then have to reshuffle one of the table's columns so that corresponding columns in the 2 tables are correctly aligned (merged). This added functionality buys very little to the user, IMO, while creating a performance sink and delaying integration. Also, this was not in the legacy code, and added functionality like that should have probably been discussed when we agreed on the API, not just before code-freeze.

Reply:
@jrhemstad, consider this: `left_key_columns = {2,4}`, `right_key_columns = {1,3}` and both key columns are the same type. Since you cannot "match" key columns based on types, how do you match them, then? 

IOW do I do a lex-merge assuming `left.columns(2)` corresponds to `right.columns(1)` or right.columns(4)`? Of course, the problem gets compounded when you have even more than 2 key columns. So, then you now need an additional map parameter that makes this mapping explicit, say, `{2 <-> 3, 4 <-> 1}`.

Reply:
Yeah, I was mistaken. Thanks for thinking it through and explaining it to me.